File: Core/src/main/java/com/hqq/core/toolbar/BaseToolBar.java
package com.hqq.core.toolbar;


import android.app.Activity;
import android.support.annotation.ColorInt;
import android.support.annotation.ColorRes;
import android.view.View;
import android.view.ViewGroup;
import android.widget.LinearLayout;

import com.hqq.core.R;
import com.hqq.core.utils.ScreenUtils;

/**
 * @Author : huangqiqiang
 * @Package : com.core.library.toolbar
 * @FileName :   BaseToolBar
 * @Date : 2018/11/15 0015  下午 7:26
 * @Descrive :
 * @Email :  qiqiang213@gmail.com
 * -  第一层 背景    toolBar
 * -  第二层 状态栏  statusBar
 * -  第三层 标题栏  TitleBar
 */
public abstract class BaseToolBar implements IToolBar {
    /**
     * mLinearLayout  包含
     * 1. 顶部状态栏背景颜色
     * 2. 中间 toolbar 与 toolbar  背景颜色  与 底部的分割线
     */

    /**
     * 根布局
     */
    public View mRootView = null;
    /**
     * 状态栏背景颜色
     */
    protected int mDefStatusColor = R.color.toolbar_status_color;

    /**
     * 状态栏背景
     */
    protected View mStatusBar;
    /**
     * 当前activity
     */
    protected Activity mActivity;
    /**
     * 底部分割线
     */
    protected View mViewLine;
    /**
     * 是否显示 分割线
     */
    protected boolean mIsShowLine = true;

    protected boolean mIsShowBar = true;

    protected boolean mIsShowStatusBar = true;

    @Override
    public BaseToolBar setShowStatusBar(boolean showStatusBar) {
        mIsShowStatusBar = showStatusBar;
        return this;
    }

    /**
     * 创建 toolbar
     */
    @Override
    public BaseToolBar createToolBar(final Activity activity) {
        mActivity = activity;
        LinearLayout linearLayout = new LinearLayout(activity);
        linearLayout.setLayoutParams(new LinearLayout.LayoutParams(ViewGroup.LayoutParams.MATCH_PARENT, ViewGroup.LayoutParams.WRAP_CONTENT));
        linearLayout.setOrientation(LinearLayout.VERTICAL);


        if (mIsShowStatusBar) {
            initStatusBar(mActivity);
            int mStatusBarHeight = ScreenUtils.getStatusBarHeight(activity);
            linearLayout.addView(mStatusBar, new LinearLayout.LayoutParams(ViewGroup.LayoutParams.MATCH_PARENT, mStatusBarHeight));
        }

        // 目前不知道是否需要都执行 createToolBar
        View toolBar = iniToolBar(activity, linearLayout);
        if (mIsShowBar && toolBar != null) {
            linearLayout.addView(toolBar);
        }

        if (mIsShowLine) {
            mViewLine = new View(mActivity);
            mViewLine.setLayoutParams(new LinearLayout.LayoutParams(ViewGroup.LayoutParams.MATCH_PARENT, ScreenUtils.dip2px(mActivity, 1)));
            mViewLine.setBackgroundResource(R.color.toolbar_line_bg);
            linearLayout.addView(mViewLine);
        }
        mRootView = linearLayout;
        return this;
    }


    /***************** statusBar  *************************/
    /**
     * 获取状态 底部的View
     *
     * @return
     */
    public View getStatusBar() {
        return mStatusBar;
    }

    /**
     * 设置 状态栏底部的颜色
     *
     * @param colorInt
     */
    @Override
    public BaseToolBar setDefStatusColor(@ColorInt int colorInt) {
        if (getStatusBar() != null) {
            getStatusBar().setBackgroundColor(colorInt);
        }else {
            mDefStatusColor=colorInt;
        }
        return this;
    }

    @Override
    public BaseToolBar setShowBar(boolean showBar) {
        mIsShowBar = showBar;
        return this;
    }

    /**
     * 滑动渐变
     *
     * @param alpha 0 -1
     */
    public void initScroll(float alpha) {
        if (mStatusBar != null) {
            mStatusBar.setAlpha(alpha);
        }
        if (mViewLine != null) {
            mViewLine.setAlpha(alpha);
        }
    }

    /**
     * 初始化状态栏  底部的view
     *
     * @param activity
     */
    public void initStatusBar(Activity activity) {
        mStatusBar = new View(activity);
        mStatusBar.setBackgroundResource(mDefStatusColor);
    }
    /***************** Toolbar  *************************/

    /**
     * 获取 根布局
     *
     * @return mRootView
     */
    @Override
    public View getRootView() {
        return mRootView;
    }

    /***************** Line  *************************/
    /**
     * 需要在 createToolBar 之前调用才会生效
     * 这个方法会导致 mViewLine 是 null
     * 但是会优化布局
     * {@link #showLine}
     *
     * @param showLine
     */
    @Deprecated
    public void setShowLine(boolean showLine) {
        mIsShowLine = showLine;
    }

    /**
     * 获取分割线
     *
     * @return mViewLine
     */
    public View getViewLine() {
        return mViewLine;
    }

    /**
     * 是否显示分割线
     *
     * @param isShow isShow
     */
    public void showLine(boolean isShow) {
        if (isShow) {
            getViewLine().setVisibility(View.VISIBLE);
        } else {
            getViewLine().setVisibility(View.GONE);
        }
    }

    /**
     * 设置分割线颜色
     *
     * @param color color id
     */
    public void setLineColor(int color) {
        getViewLine().setBackgroundResource(color);
    }

}

File: Core/src/main/java/com/hqq/core/toolbar/BaseToolBar.kt
package com.hqq.core.toolbar

import android.app.Activity
import android.view.View
import android.view.ViewGroup
import android.widget.LinearLayout
import androidx.annotation.ColorInt
import androidx.annotation.ColorRes
import com.hqq.core.R
import com.hqq.core.utils.ScreenUtils
import java.lang.ref.WeakReference

/**
 * @Author : huangqiqiang
 * @Package : com.core.library.toolbar
 * @FileName :   BaseToolBar
 * @Date : 2018/11/15 0015  下午 7:26
 * @Descrive :
 * @Email :  qiqiang213@gmail.com
 * -  第一层 背景    toolBar
 * -  第二层 状态栏  statusBar
 * -  第三层 标题栏  TitleBar
 */
abstract class BaseToolBar : IToolBar {
    /**
     * mLinearLayout  包含
     * 1. 顶部状态栏背景颜色
     * 2. 中间 toolbar 与 toolbar  背景颜色  与 底部的分割线
     */
    /**
     * 获取 根布局
     *
     * @return mRootView
     */
    /**
     * 根布局
     */
    override var rootView: View? = null

    /**
     * 状态栏背景颜色
     */
    @ColorRes
    protected var defStatusColor = R.color.toolbar_status_color

    /**
     * 获取状态 底部的View
     *
     * @return
     */
    /**
     * 状态栏背景
     */
    var statusBar: View? = null
        protected set

    /**
     * 当前activity
     */
    protected var activity: WeakReference<Activity?>? = null

    /**
     * 获取分割线
     *
     * @return mViewLine
     */
    /**
     * 底部分割线
     */
    var viewLine: View? = null
        protected set

    /**
     * 是否显示 分割线
     */
    protected var isShowLine = true

    /**
     * 是否显示 标题栏
     */
    protected var isShowBar = true

    /**
     * 是否显示状态栏
     */
    protected var isShowStatusBar = true

    override fun setShowStatusBar(showStatusBar: Boolean): BaseToolBar {
        isShowStatusBar = showStatusBar
        return this
    }

    /**
     * 创建 toolbar
     */
    override fun createToolBar(activity: Activity?): BaseToolBar {
        this.activity = WeakReference(activity)
        val linearLayout = LinearLayout(activity)
        linearLayout.layoutParams = LinearLayout.LayoutParams(ViewGroup.LayoutParams.MATCH_PARENT, ViewGroup.LayoutParams.WRAP_CONTENT)
        linearLayout.orientation = LinearLayout.VERTICAL
        if (isShowStatusBar) {
            initStatusBar(this.activity!!.get())
            val mStatusBarHeight = ScreenUtils.getStatusBarHeight4Resources(activity)
            linearLayout.addView(statusBar, LinearLayout.LayoutParams(ViewGroup.LayoutParams.MATCH_PARENT, mStatusBarHeight))
        }

        // 目前不知道是否需要都执行 createToolBar
        val toolBar = iniToolBar(activity!!, linearLayout)
        if (isShowBar && toolBar != null) {
            linearLayout.addView(toolBar)
        }
        if (isShowLine) {
            viewLine = View(this.activity!!.get())
            viewLine!!.layoutParams = LinearLayout.LayoutParams(ViewGroup.LayoutParams.MATCH_PARENT, ScreenUtils.dip2px(this.activity!!.get(), 1f))
            viewLine!!.setBackgroundResource(R.color.toolbar_line_bg)
            linearLayout.addView(viewLine)
        }
        rootView = linearLayout
        return this
    }
    /***************** statusBar   */

    /**
     * 设置 状态栏底部的颜色
     *
     * @param colorInt
     */
    override fun setDefStatusColor(@ColorRes colorInt: Int): BaseToolBar {
        if (statusBar != null) {
            statusBar!!.setBackgroundResource(colorInt)
        } else {
            defStatusColor = colorInt
        }
        return this
    }

    override fun setShowBar(showBar: Boolean): BaseToolBar {
        isShowBar = showBar
        return this
    }

    /**
     * 滑动渐变
     *
     * @param alpha 0 -1
     */
    open fun initScroll(alpha: Float) {
        if (statusBar != null) {
            statusBar!!.alpha = alpha
        }
        if (viewLine != null) {
            viewLine!!.alpha = alpha
        }
    }

    /**
     * 初始化状态栏  底部的view
     *
     * @param activity
     */
    fun initStatusBar(activity: Activity?) {
        statusBar = View(activity)
        statusBar!!.setBackgroundResource(defStatusColor)
    }
    /***************** Toolbar   */
    /***************** Line   */


    /**
     * 是否显示分割线
     *
     * @param isShow isShow
     */
    fun showLine(isShow: Boolean) {
        if (isShow) {
            viewLine!!.visibility = View.VISIBLE
        } else {
            viewLine!!.visibility = View.GONE
        }
    }

    /**
     * 设置分割线颜色
     *
     * @param color color id
     */
    fun setLineColor(color: Int) {
        viewLine!!.setBackgroundResource(color)
    }
}